apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -568,104 +536,183 @@ session_create(SessionId, Props) ->
+                NDSSubExt = export_subscription(NDSSub),
                 ?tp(
                     ds_session_subscription_present,
-                    #{iterator => NIterator, session_id => DSSessionId}
+                    #{sub => NDSSubExt, session_id => DSSessionId}
                 ),
-                {ok, NIterator, _IsNew = false}
+                {ok, NDSSubExt, _IsNew = false}
         end
     end).
 
-session_insert_iterator(DSSessionId, TopicFilter, Props) ->
-    {IteratorId, StartMS} = new_iterator_id(DSSessionId),
-    IteratorRef = #iterator_ref{
-        ref_id = {DSSessionId, TopicFilter},
-        it_id = IteratorId,
+-spec session_insert_subscription(id(), topic_filter(), map()) -> ds_sub().
+session_insert_subscription(DSSessionId, TopicFilter, Props) ->
+    {DSSubId, StartMS} = new_subscription_id(DSSessionId, TopicFilter),
+    DSSub = #ds_sub{
+        id = DSSubId,
         start_time = StartMS,
-        props = Props
+        props = Props,
+        extra = #{}
     },
-    ok = mnesia:write(?ITERATOR_REF_TAB, IteratorRef, write),
-    IteratorRef.
-
-session_update_iterator(IteratorRef, Props) ->
-    NIteratorRef = IteratorRef#iterator_ref{props = Props},
-    ok = mnesia:write(?ITERATOR_REF_TAB, NIteratorRef, write),
-    NIteratorRef.
-
-%% @doc Called when a client unsubscribes from a topic.
--spec session_del_iterator(id(), topic_filter()) -> ok.
-session_del_iterator(DSSessionId, TopicFilter) ->
-    IteratorRefId = {DSSessionId, TopicFilter},
+    ok = mnesia:write(?SESSION_SUBSCRIPTIONS_TAB, DSSub, write),
+    DSSub.
+
+-spec session_update_subscription(ds_sub(), map()) -> ds_sub().
+session_update_subscription(DSSub, Props) ->
+    NDSSub = DSSub#ds_sub{props = Props},
+    ok = mnesia:write(?SESSION_SUBSCRIPTIONS_TAB, NDSSub, write),
+    NDSSub.
+
+session_del_subscription(DSSessionId, TopicFilter) ->
+    DSSubId = {DSSessionId, TopicFilter},
     transaction(fun() ->
-        mnesia:delete(?ITERATOR_REF_TAB, IteratorRefId, write)
+        mnesia:delete(?SESSION_SUBSCRIPTIONS_TAB, DSSubId, write)
     end).
 
-session_del_iterator(#iterator_ref{ref_id = IteratorRefId}) ->
-    mnesia:delete(?ITERATOR_REF_TAB, IteratorRefId, write).
+session_del_subscription(#ds_sub{id = DSSubId}) ->
+    mnesia:delete(?SESSION_SUBSCRIPTIONS_TAB, DSSubId, write).
 
-session_read_iterators(DSSessionId) ->
-    % NOTE: somewhat convoluted way to trick dialyzer
-    Pat = erlang:make_tuple(record_info(size, iterator_ref), '_', [
-        {1, iterator_ref},
-        {#iterator_ref.ref_id, {DSSessionId, '_'}}
-    ]),
-    mnesia:match_object(?ITERATOR_REF_TAB, Pat, read).
+session_read_subscriptions(DSSessionId) ->
+    MS = ets:fun2ms(
+        fun(Sub = #ds_sub{id = {Sess, _}}) when Sess =:= DSSessionId ->
+            Sub
+        end
+    ),
+    mnesia:select(?SESSION_SUBSCRIPTIONS_TAB, MS, read).
+
+-spec new_subscription_id(id(), topic_filter()) -> {subscription_id(), integer()}.
```

Comment:
Nit: I think the previous return type was a bit more precise.

```suggestion
-spec new_subscription_id(id(), topic_filter()) -> {subscription_id(), emqx_ds:time()}.
```

Reply:
`emqx_ds:time()` specifies unit as microsecond, but here we need millisecond.

Reply:
Oops, my memory failed me :see_no_evil: 
Ignore me.